### DISCUSSION TITLE
VIPS Performance for HEIC vs JPEG

### DISCUSSION THREAD

**Discussion by ksylvest** [libvips/libvips#2821] [General]
I'm noticing `vipsthumbnail` is ~5 times slower when processing .heic images. I am benchmarking with:

[HEIC ~1.3 MB](https://ksylvest.s3.amazonaws.com/demo.heic)
[JPEG ~8.6 MB](https://ksylvest.s3.amazonaws.com/demo.jpeg)

```sh
time vipsthumbnail --size=800x800 demo.heic -o demo-heic.jpg[Q=80,strip] # 1.01 secs
time vipsthumbnail --size=800x800 demo.jpeg -o demo-jpeg.jpg[Q=80,strip] # 0.19 secs
```

Can anything be done to bring the thumbnail generation closer?

**Comment by jcupitt** [comment]
Hi @ksylvest,

I had a look at your test image, and it's been encoded without a thumbnail. Perhaps the writer was keen to reduce size as much as possible? This means that libvips has to decode the whole of the main image (3000 x 4000 pixels), then shrink that. It's always going to be slow, I'm afraid.

If there's an embedded thumbnail, it's closer, eg.:

```
# your image
$ time vipsthumbnail demo.heic 
real	0m0.496s
user	0m0.816s
sys	0m0.097s

# sample from an iphone12
$ time vipsthumbnail iphone12.heic 
real	0m0.065s
user	0m0.065s
sys	0m0.026s
```

Though 0.5s to decode a large image is not so bad, I think.

Of course, the embedded thumbnail needs to be larger than the output image you are asking for. An 800x800 output will almost always force decompression of the main image.

**Reply by ksylvest** [reply]
Thanks @jcupitt for the response (and thanks for all your contributions to this library!). I wasn't aware of encoding with thumbnails. Is the reason the JPEG is much faster due to something similar?

Thanks,

Kevin

**Reply by jcupitt** [reply]
I think it's a combination of factors.  First, JPEG is pretty old now, so very clever engineers have spent years tuning the libraries. And proably more importantly, libjpeg supports shrink-on-load, which can give a huge boost to thumbnailing performance.

When you decode a JPEG, you can set a subsample factor of 1, 2, 4 or 8, and libjpeg will only decode enough of the coefficients to generate  an image of that size. 8x shrink is especially quick since it just needs to fetch the DC coefficient of each DCT block.

You can see the difference like this:

```
# thumbnail_image can't use shrink-on-load
$ time vips thumbnail_image nina.jpg x.jpg 128

real	0m0.347s
user	0m0.289s
sys	0m0.108s

# but straight thumbnail can
$ time vips thumbnail nina.jpg x.jpg 128

real	0m0.058s
user	0m0.047s
sys	0m0.012s
```
